### DISCUSSION TITLE
How to read configMap from cluster and use those values to create the component?

### DISCUSSION THREAD

**Discussion by miguelslemos** [kubevela/kubevela#6384] [Q&A]
Hi! I would like to create a component using values from configMap inside the cluster. How can I do it?.

is the only way using workflow-step?

**Comment by miguelslemos** [comment]
I could read a cluster configMap and pass this information using the read-object operation. However, the aim is to make this information available to all components/traits. So it would have to be something more generic.

It's like injecting a context from the cluster like: aws-account-id, stack-id, etc.

**Comment by miguelslemos** [comment]
c/c @chivalryq sorry to mention you, but could you help here, please?

**Comment by FogDong** [comment]
Hi, you can checkout this doc, note that the action is only valid in workflowsteps.
https://kubevela.io/docs/platform-engineers/workflow/cue-actions/#dovar

**Comment by miguelslemos** [comment]
Thanks for your reply. I want to make the application as simple as possible for end users.

In the suggestion you commented, is it possible to create a custom workflowstep that reads an existing configMap using op.#Read and uses op.ws.#DoVar to add the configMap data in the context? Therefore, it is possible to use the configMap data in the custom component using the context variable

**Reply by FogDong** [reply]
Yes. The process will be:
1. Write a custom workflow step to inject var from configMap.
2. Write a custom workflow step to use the context var to apply component.

**Comment by oanasc** [comment]
If I might add to this discussion, was looking for a similar solution as we're starting to adopt KubeVela at my company.
The custom context that is available on [workflowRun](https://kubevela.io/docs/end-user/pipeline/workflowrun/#custom-context-data) would be ideal. Sadly I don't see the same supported in the Application/Workflow.
My intention was to have a generic way of injecting all the extra variables that we might need such as image tag, other metadata and keep the application definition simple and free of env variables, then in the cue templates use the context with all the injected variables and use them to fill in any gaps, such as append the docker image tag. So would be very beneficial in my opinion having custom contexts capabilities.
Very interested in any other ways around it or please correct me if I've got this wrong.

**Reply by FogDong** [reply]
Exactly. The first solution comes to me is also the context in workflowRun. But unfortunately this feature is not exposed to Application Workflow right now. If this is a feature that the community want, please open a feature request.

**Reply by oanasc** [reply]
Thanks @FogDong  For anyone stumbling on this chat here is a link to the feature request: https://github.com/kubevela/kubevela/issues/6391
